### PR TITLE
consensus/ethash, params: implement eip-2384: bump difficulty bomb

### PR DESCRIPTION
**Pull request by holiman**
WIP, this implements https://github.com/ethereum/EIPs/pull/2384, which nudges the difficulty bomb another 4M blocks (9M in total). 
cc @shemnon

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -293,6 +293,7 @@ type ChainConfig struct {
 	PetersburgBlock     *big.Int `json:"petersburgBlock,omitempty"`     // Petersburg switch block (nil = same as Constantinople)
 	IstanbulBlock       *big.Int `json:"istanbulBlock,omitempty"`       // Istanbul switch block (nil = no fork, 0 = already on istanbul)
 	EWASMBlock          *big.Int `json:"ewasmBlock,omitempty"`          // EWASM switch block (nil = no fork, 0 = already activated)
+	EIP2384Block        *big.Int `json:"eip2384Block,omitempty"`        // Eip-2384 (bomb delay) switch block (nil = no fork, 0 = already activated)
```
Comment (karalabe) [inline params/config.go]: Nitpick, please move this above EWASM, since that's an experimental feature only.

File: consensus/ethash/consensus.go [hunk 2]
```diff
@@ -44,6 +44,12 @@ var (
 	maxUncles                 = 2                 // Maximum number of uncles allowed in a single block
 	allowedFutureBlockTime    = 15 * time.Second  // Max time from current time allowed for blocks, before they're considered future blocks
 
+	// calcDifficultyEip2384 is the difficulty adjustment algorithm as specified by EIP 2384.
+	// It offsets the bomb 4M blocks from Constantinople, so in total 9M blocks.
+	// Specification EIP-2384 :https://github.com/ethereum/EIPs/pull/2384
+	// TODO: change link once it's merged
```
Comment (karalabe) [inline consensus/ethash/consensus.go]: https://eips.ethereum.org/EIPS/eip-2384

File: params/config.go [hunk 3]
```diff
@@ -380,6 +382,10 @@ func (c *ChainConfig) IsConstantinople(num *big.Int) bool {
 	return isForked(c.ConstantinopleBlock, num)
 }
 
+func (c *ChainConfig) IsEip2384(num *big.Int) bool {
```
Comment (karalabe) [inline params/config.go]: Please add a method doc.

**Comment by holiman** [comment]
Updated with block numbers, and I changed the genesis name to `muirGlacierBlock`

File: params/config.go [hunk 4]
```diff
@@ -380,6 +384,11 @@ func (c *ChainConfig) IsConstantinople(num *big.Int) bool {
 	return isForked(c.ConstantinopleBlock, num)
 }
 
+// IsEip2384 returns whether num is equal or greater than the EIP 2384 block
+func (c *ChainConfig) IsEip2384(num *big.Int) bool {
```
Comment (karalabe) [inline params/config.go]: Nit, please make `Eip` all caps.

File: params/config.go [hunk 5]
```diff
@@ -380,6 +384,11 @@ func (c *ChainConfig) IsConstantinople(num *big.Int) bool {
 	return isForked(c.ConstantinopleBlock, num)
 }
 
+// IsEip2384 returns whether num is equal or greater than the EIP 2384 block
```
Comment (karalabe) [inline params/config.go]: Please copy paste the doc form a fork from above to have it consistent across all.

File: params/config.go (hunk 4 above)
Comment (karalabe) [reply params/config.go]: No, please rename to IsMuriGlacier

File: core/forkid/forkid_test.go [hunk 6]
```diff
@@ -57,8 +57,10 @@ func TestCreation(t *testing.T) {
 				{7279999, ID{Hash: checksumToBytes(0xa00bc324), Next: 7280000}}, // Last Byzantium block
 				{7280000, ID{Hash: checksumToBytes(0x668db0af), Next: 9069000}}, // First and last Constantinople, first Petersburg block
 				{9068999, ID{Hash: checksumToBytes(0x668db0af), Next: 9069000}}, // Last Petersburg block
-				{9069000, ID{Hash: checksumToBytes(0x879d6e30), Next: 0}},       // Today Istanbul block
-				{10000000, ID{Hash: checksumToBytes(0x879d6e30), Next: 0}},      // Future Istanbul block
+				{9069000, ID{Hash: checksumToBytes(0x879d6e30), Next: 9200000}}, // Muir Glacier block
+				{9199999, ID{Hash: checksumToBytes(0x879d6e30), Next: 9200000}}, // Muir Glacier block
+				{9200000, ID{Hash: checksumToBytes(0xe029e991), Next: 0}},       // Future Istanbul block
+				{10000000, ID{Hash: checksumToBytes(0xe029e991), Next: 0}},      // Future Istanbul block
```
Comment (karalabe) [inline core/forkid/forkid_test.go]: Comments are off

File: params/config.go [hunk 7]
```diff
@@ -292,6 +294,7 @@ type ChainConfig struct {
 	ConstantinopleBlock *big.Int `json:"constantinopleBlock,omitempty"` // Constantinople switch block (nil = no fork, 0 = already activated)
 	PetersburgBlock     *big.Int `json:"petersburgBlock,omitempty"`     // Petersburg switch block (nil = same as Constantinople)
 	IstanbulBlock       *big.Int `json:"istanbulBlock,omitempty"`       // Istanbul switch block (nil = no fork, 0 = already on istanbul)
+	EIP2384Block        *big.Int `json:"muirGlacierBlock,omitempty"`    // Eip-2384 (bomb delay) switch block (nil = no fork, 0 = already activated)
```
Comment (karalabe) [inline params/config.go]: Please change to MuirGlacierBlock

File: params/config.go [hunk 8]
```diff
@@ -432,6 +441,7 @@ func (c *ChainConfig) CheckConfigForkOrder() error {
 		{"constantinopleBlock", c.ConstantinopleBlock},
 		{"petersburgBlock", c.PetersburgBlock},
 		{"istanbulBlock", c.IstanbulBlock},
+		{"eip2384Block", c.EIP2384Block},
```
Comment (karalabe) [inline params/config.go]: Please fix name, both sides actually.

File: params/config.go [hunk 9]
```diff
@@ -485,6 +495,9 @@ func (c *ChainConfig) checkCompatible(newcfg *ChainConfig, head *big.Int) *Confi
 	if isForkIncompatible(c.IstanbulBlock, newcfg.IstanbulBlock, head) {
 		return newCompatError("Istanbul fork block", c.IstanbulBlock, newcfg.IstanbulBlock)
 	}
+	if isForkIncompatible(c.EIP2384Block, newcfg.EIP2384Block, head) {
+		return newCompatError("EIP-2348 fork block", c.EIP2384Block, newcfg.EIP2384Block)
```
Comment (karalabe) [inline params/config.go]: Please change the error message to Muir Glacier

**Comment by holiman** [comment]
I now added an override aswell
```
[user@work go-ethereum]$ build/bin/geth --override.muirglacier=100
INFO [12-06|09:39:50.223] Bumping default cache on mainnet         provided=1024 updated=4096
WARN [12-06|09:39:50.223] Sanitizing cache to Go's GC limits       provided=4096 updated=2589
INFO [12-06|09:39:50.225] Maximum peer count                       ETH=50 LES=0 total=50
INFO [12-06|09:39:50.225] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
ERROR[12-06|09:39:50.253] Failed to enumerate USB devices          hub=ledger vendor=11415 failcount=1 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:39:50.253] Failed to enumerate USB devices          hub=trezor vendor=21324 failcount=1 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:39:50.253] Failed to enumerate USB devices          hub=trezor vendor=4617  failcount=1 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:39:50.253] Failed to enumerate USB devices          hub=ledger vendor=11415 failcount=2 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:39:50.253] Failed to enumerate USB devices          hub=trezor vendor=21324 failcount=2 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:39:50.253] Failed to enumerate USB devices          hub=trezor vendor=4617  failcount=2 err="failed to initialize libusb: libusb: unknown error [code -99]"
INFO [12-06|09:39:50.256] Starting peer-to-peer node               instance=Geth/v1.9.8-unstable-51172f0a-20191205/linux-amd64/go1.13.4
INFO [12-06|09:39:50.257] Allocated trie memory caches             clean=647.00MiB dirty=647.00MiB
INFO [12-06|09:39:50.257] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=1.26GiB handles=262144
INFO [12-06|09:39:50.722] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient
Fatal: Error starting protocol stack: unsupported fork ordering: istanbulBlock enabled at 9069000, but muirGlacierBlock enabled at 100
[user@work go-ethereum]$ build/bin/geth --override.istanbul=99 --override.muirglacier=100
INFO [12-06|09:40:12.680] Bumping default cache on mainnet         provided=1024 updated=4096
WARN [12-06|09:40:12.680] Sanitizing cache to Go's GC limits       provided=4096 updated=2589
INFO [12-06|09:40:12.682] Maximum peer count                       ETH=50 LES=0 total=50
INFO [12-06|09:40:12.682] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
ERROR[12-06|09:40:12.683] Failed to enumerate USB devices          hub=ledger vendor=11415 failcount=1 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:40:12.683] Failed to enumerate USB devices          hub=trezor vendor=21324 failcount=1 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:40:12.683] Failed to enumerate USB devices          hub=trezor vendor=4617  failcount=1 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:40:12.684] Failed to enumerate USB devices          hub=ledger vendor=11415 failcount=2 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:40:12.684] Failed to enumerate USB devices          hub=trezor vendor=21324 failcount=2 err="failed to initialize libusb: libusb: unknown error [code -99]"
ERROR[12-06|09:40:12.684] Failed to enumerate USB devices          hub=trezor vendor=4617  failcount=2 err="failed to initialize libusb: libusb: unknown error [code -99]"
INFO [12-06|09:40:12.684] Starting peer-to-peer node               instance=Geth/v1.9.8-unstable-51172f0a-20191205/linux-amd64/go1.13.4
INFO [12-06|09:40:12.684] Allocated trie memory caches             clean=647.00MiB dirty=647.00MiB
INFO [12-06|09:40:12.684] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=1.26GiB handles=262144
INFO [12-06|09:40:12.714] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient
Fatal: Error starting protocol stack: unsupported fork ordering: petersburgBlock enabled at 7280000, but istanbulBlock enabled at 99
[user@work go-ethereum]$ 
```

File: params/config.go [hunk 10]
```diff
@@ -485,6 +495,9 @@ func (c *ChainConfig) checkCompatible(newcfg *ChainConfig, head *big.Int) *Confi
 	if isForkIncompatible(c.IstanbulBlock, newcfg.IstanbulBlock, head) {
 		return newCompatError("Istanbul fork block", c.IstanbulBlock, newcfg.IstanbulBlock)
 	}
+	if isForkIncompatible(c.MuirGlacierBlock, newcfg.MuirGlacierBlock, head) {
+		return newCompatError("EIP-2348 fork block", c.MuirGlacierBlock, newcfg.MuirGlacierBlock)
```
Comment (karalabe) [inline params/config.go]: EIP-2348 -> Muir Glacier

**Comment by currystar2** [comment]
https://documentcloud.adobe.com/link/track?uri=urn%3Aaaid%3Ascds%3AUS%3A3e4af3d1-6560-4dad-8f51-98b8f64efea8

**Pull request merged**